### ISSUE TITLE
Failure When Attempting to Deploy via ARM 'Unable to resolve Azure Files Settings from Key Vault. Unable to resolve setting: WEBSITE_CONTENTAZUREFILECONNECTIONSTRING'

### ISSUE THREAD

**Issue by aolszowka** [Azure/azure-functions-host#8992]
This issue is similar, but not identical to an existing open reported issue #7094 specifically the comment [from @haroldwongms](https://github.com/Azure/azure-functions-host/issues/7094#issuecomment-769076742) in their comments they indicate that they resolved the issue but didn't describe what the root cause or what their fix was.

Consider the following truncated ARM Template resources:

```json
{
    "type": "Microsoft.KeyVault/vaults/secrets",
    "apiVersion": "2022-07-01",
    "name": "[concat(variables('keyVaultName'), '/stg-conn-str')]",
    "comments": "This calculated value is used to store the connection string to the Storage Account which is used by the Function Application",
    "dependsOn": [
        "[resourceId('Microsoft.KeyVault/vaults', variables('keyVaultName'))]",
        "[resourceId('Microsoft.Storage/storageAccounts', variables('storageAccountName'))]"
    ],
    "properties": {
        "value": "[concat('DefaultEndpointsProtocol=https;AccountName=', variables('storageAccountName'), ';AccountKey=', listKeys(variables('storageAccountName'),'2022-05-01').keys[0].value)]"
    }
},
{
    "type": "Microsoft.Web/sites/config",
    "apiVersion": "2022-03-01",
    "name": "[concat(variables('functionAppName'), '/appsettings')]",
    "dependsOn": [
        "[resourceId('microsoft.insights/components', variables('applicationInsightsName'))]",
        "[resourceId('Microsoft.KeyVault/vaults', variables('keyVaultName'))]",
        "[resourceId('Microsoft.KeyVault/vaults/secrets', variables('keyVaultName'), 'someothersecret')]",
        "[resourceId('Microsoft.KeyVault/vaults/secrets', variables('keyVaultName'), 'stg-conn-str')]",
        "[resourceId('Microsoft.Web/sites', variables('functionAppName'))]"
    ],
    "properties": {
        "APPLICATIONINSIGHTS_CONNECTION_STRING": "[reference(resourceId('microsoft.insights/components/', variables('applicationInsightsName'))).ConnectionString]",
        "AzureWebJobsStorage": "[concat('@Microsoft.KeyVault(SecretUri=', reference(resourceId('Microsoft.KeyVault/vaults/secrets', variables('keyVaultName'), 'stg-conn-str')).secretUri, ')')]",
        "someothersecret": "[concat('@Microsoft.KeyVault(SecretUri=', reference(resourceId('Microsoft.KeyVault/vaults/secrets', variables('keyVaultName'), 'someothersecret')).secretUri, ')')]",
        "FUNCTIONS_EXTENSION_VERSION": "~4",
        "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING": "[concat('@Microsoft.KeyVault(SecretUri=', reference(resourceId('Microsoft.KeyVault/vaults/secrets', variables('keyVaultName'), 'stg-conn-str')).secretUri, ')')]",
        "WEBSITE_CONTENTSHARE": "[tolower(variables('functionAppName'))]",
        "WEBSITE_ENABLE_SYNC_UPDATE_SITE": "true",
        "WEBSITE_RUN_FROM_PACKAGE": "1"
    }
}
```

This fails intermittently in a deployment with the error:

```json
{
  "code": "DeploymentFailed",
  "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.",
  "details": [
    {
      "message": "Unable to resolve Azure Files Settings from Key Vault. Details: Unable to resolve setting: WEBSITE_CONTENTAZUREFILECONNECTIONSTRING with error: AccessToKeyVaultDenied."
    }
  ]
}
```

Redeployment of the same code works in some scenarios which is frustrating.

#### Investigative information
I have already confirmed that the KeyVault being Accessed is correctly setup to allow reads from the deployment process; other reads (such as `someothersecret`) work without issue which to me rules out issues with the Key Vault permissions, the fact that this works on redeploy from time to time is also an indication to me that something else is at play here.

It really seems to have a problem with `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` which is confusing because `AzureWebJobsStorage` is set identically.

Speculation says that this maybe a timing issue; however, I explicitly have the `dependsOn` so I believe I have done all I can to tell the template that I NEED the secret created BEFORE initializing this.

#### Repro steps
Attempt to perform a deployment that utilizes a secret like the above. Intermittently you will encounter the error.

If it would help, I can work on a more elaborate ARM template that should minimally reproduce this issue, but I need to know that this effort will be taken seriously.

#### Expected behavior
This should deploy in all scenarios without issue.

#### Actual behavior
Intermittently you will encounter the following error:

```json
{
  "code": "DeploymentFailed",
  "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.",
  "details": [
    {
      "message": "Unable to resolve Azure Files Settings from Key Vault. Details: Unable to resolve setting: WEBSITE_CONTENTAZUREFILECONNECTIONSTRING with error: AccessToKeyVaultDenied."
    }
  ]
}
```

#### Known workarounds
A redeployment of this will usually work; however, this is not great for putting this into a reliable CI/CD pipeline.

Previous to attempting to use a KeyVault reference I was simply hardcoding in the value 

#### Related information 
We have an active Azure Support contract, and I will open up a Sev C case on that end as well, pointing back to this issue.

**EDIT** open Support Request 2212190010002007

@haroldwongms sorry to ping you; but did you end up with a solution? or at least more research that might help troubleshoot this?

**Comment by aolszowka** [comment]
Microsoft Support got back to me very quickly with some possible root causes:

_START MICROSOFT SUPPORT RECOMMENDATION_

**Problem Statement:**
  
Unable to resolve Azure Files Settings from Key Vault. Details: Unable to resolve setting: WEBSITE_CONTENTAZUREFILECONNECTIONSTRING with error: AccessToKeyVaultDenied.
   
**Observation/Findings:**
  
This issue can happen due to a few different reasons and is mentioned in our documentation for using Key Vault References.
 
[Use Key Vault references - Azure App Service | Microsoft Learn](https://learn.microsoft.com/en-us/azure/app-service/app-service-key-vault-references?tabs=azure-cli#considerations-for-azure-files-mounting)
 
As you state that this usually succeeds on a second attempt, this is likely due to the location of your app settings in your template. If the settings are not created along with the function app resource, the function app is created using the default file system and will cause this to fail. On a second attempt, it will switch the file system and succeed. More information on this can be found on our project kudu wiki page linked below.
 
[WEBSITE_RUN_FROM_PACKAGE and WEBSITE_CONTENTAZUREFILECONNECTIONSTRING Best Practices · projectkudu/kudu Wiki (github.com)](https://github.com/projectkudu/kudu/wiki/WEBSITE_RUN_FROM_PACKAGE-and-WEBSITE_CONTENTAZUREFILECONNECTIONSTRING-Best-Practices#best-practice)
  
**Action Plan:**
  
Please review the documentation provided on the WEBSITE_CONTETNAZUREFILECONNECTIONSTRING setting. You may need to adjust your template to include your app settings in the function resource as suggested in the project kudu wiki.

_END MICROSOFT SUPPORT RECOMMENDATION_

For me I believe I correctly understood what was said in the linked articles and checked my ARM Templates, however I'm still encountering failures and will update this thread once a solution is reached. I wanted to share the above in case that was someone else's root cause/issue.

**Comment by aolszowka** [comment]
With help of the Microsoft Support Agent (@v-lhoffmann) we were able to narrow down the root cause:

_It looks like your app settings are being created before the access policies for the resource are created. The reason this works on a second try, is that the policy is already created and is able to apply the app settings._
 
_To work around this, you would want to make sure that the policies are in place before the app settings or split the template to add all of the resources first, then add the settings. Since the resource manager tries to do everything in parallel, some things will finish before others._

In hindsight the issue would have been more obvious had I included a few more portions of the ARM template above to reproduce the issue. This was an interesting combination of the behavior of ARM Deployments (performing resource deployments in Parallel) the use of a managed identity for the Function Service, and the order in which the access policies to the KeyVault were setup. The intermittent success was a race condition depending on the deployment time.

I have opened https://github.com/MicrosoftDocs/azure-docs/issues/103042 to try and improve the documentation for this specific scenario as well as provided a more in-depth ARM template that would show the issue as well as further analysis. For now, I am going to close this as resolved with the hope that it helps someone else in the future.
